### DISCUSSION TITLE
Making documentation easy

### DISCUSSION THREAD

**Discussion by bclothier** [twinbasic/twinbasic#597] [Ideas]
Often, documentation comes as an afterthought and if ever done, is usually low-quality and doesn't cover everything. The bigger problem, however, is keeping documentation in synchronization with the actual behavior. Another major problem is that documentation are usually provided as text, which hamper using screenshots or other non-textual input to help supplement the documentation.

Just to be clear, I'm speaking primarily about documentation for other developers, rather than for the end users.

In Rust ecosystem, this is solved by making documentation a part of the crate. [Here's an example](https://docs.rs/syn/1.0.83/syn/). By providing the `docs.rs` and `crates.io`, Rust make it very easy to find & download external libraries (see #16 ) but also provide an uniform documentation system that everyone can use which is a large advantage. The less time I spend learning about a documentation's formatting, the more productive I am on using the documentation itself. 

In Python, [DocTest](https://www.digitalocean.com/community/tutorials/how-to-write-doctests-in-python) provides documentation that can be tested & verified by code, which really helps in the area of "don't tell me, show me". Because that can be verified, it makes it easy to assert that the documented behavior is in synchronization with the implementation.

In .NET, we have [xmldocs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/) which also have tools (e.g. DocFx, SandCastle, Doxygen) for generating help files from the XML comments. 

VBx already has some support (e.g. `HelpFile` and `HelpContextID`) but I suspect they are universally ignored because it's usually extra work, and is "out of sight, out of mind". To compound this, MS no longer maintain the help wizard (there are multiple incarnations, I can't remember which is latest but even the latest is no longer actively supported and is hard to find). 

In order for tB to succeed, documentation has to be a first-class citizen, is easy to do, and can provide easy way to test and verify the documentation. We have the testing API which can help contribute somehow to the documentation. That could be where CodeLen feature helps by referencing the tests. Because we will have a GUI, it seems helpful to be able to provide images to help document the GUI aspect. One possible implementation is to provide a rich editor that can take a screenshot and be pasted into the text and then saved into the `.twinproj` as a supplement to the corresponding `.twin` file and again use CodeLen to show/expand the rich documentation.

**Comment by wqweto** [comment]
![image](https://user-images.githubusercontent.com/987963/147251489-f173e14b-dc8e-4ab3-a973-8607c4a12f4d.png)

**Reply by mburns08109** [reply]
OMG! - the PERFECT illustration of BAD documentation! (You have to open the box to see the instruction to open the box!?? really!?)

**Reply by wqweto** [reply]
I though the text is on the outside :-))

Nevertheless it's still absolutely useless like in *"Method Move moves the button to some desired postion"* kind of documentation of which I've seen a lot.

Although must admit rust made significant progress on the subject. They are hosting the docs centralized for free as well which makes everything discoverable.

**Reply by bclothier** [reply]
>  Nevertheless it's still absolutely useless like in "Method Move moves the button to some desired postion" kind of documentation of which I've seen a lot.

I agree wholeheartedly. I have seen some attempt to automate the documentation which always ends as a resounding failure. That’s why I cited DocTests because documentation that is testable is a verifiable documentation which avoids this type of situations:

```
‘Take a string that conatins a number
Public Function DoIt(SomeNumber As Long) As Long
```

Mr. Burns make an excellent point regarding errors. That’s another lesson we can take from Rust:

[`Option` as described in the book](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html)
[An example of error listed for opening a file](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html)

Many times I have wished that VBx documentation were more precise when it coms to handling variants. Several times, the documentation fail to note that it can only accept a subset of subtypes of variants (a common problem subtype is the `Null` AKA `VT_NULL`; several variant-accepting functions throw an error for such input).

Another form of documentation is a flow chart showing dependencies or relationship between objects. That can be very useful in seeing how you can get to B from A without having to trawl through a list of methods. 

Finally, nobody has truly succeeded in documenting the _why_ into the code. We can help document the _how_ by using good names for tokens but not the _why_.

Those are pretty hard to do well, and I argue that if the language doesn’t make it easy to do this, nobody will bother documenting their works and we end up with 100 implementations doing the same thing (mostly) because they each decided that reinventing the wheel was easier than to try and read someone else’s quirky implementation.

**Comment by mburns08109** [comment]
I will reiterate my documentation wishlist item from long ago, that I mentioned previously.

I have always wished for a Generated Errors List for any documented method/function call.
This would allow the caller of the function to anticipate the likely errors and defensively code-in handlers for those conditions.

Explanation:
if Foo() can generate application-specific errors internal to its own implementation (as opposed to those errors it can "inherit" from calls it makes to other subs/functions/methods) - Think: it uses Throw() (or err.raise in VBx terms) to generate the error itself here - then decent documentation for foo() should include a list of those errors - both the number and description of the error (as well as any developer's note on the error for preferred resolution, etc.).

Without any such list - coding error handlers is pretty much left to guesswork and/or "experience", which is a sucky position to be in for a developer/coder.
My all-time favorite example for this is the profoundly lacking documentation for (DAO or ADO - both apply here, so take your pick) OpenRecordset() - with the varying parameters and which other parameter values are valid with one another or not, a list of the error numbers for these methods would GREATLY enhance programmability - especially for coders new to using the function.

**Comment by mwolfe02** [comment]
I'm a huge fan of Python's Doc Tests, so much so that I [implemented them in VBA](https://nolongerset.com/python-inspired-doc-tests-in-vba/).  Of course, my implementation required some ~~kludgy~~ clever hacks and workarounds.  I'd love to see tB provide native support for doc tests at the language level.
